examples/vae/train_vae.py

```diff
@@ -0,0 +1,548 @@
+        args.lr_scheduler,
+        optimizer=optimizer,
+        num_warmup_steps=args.lr_warmup_steps * args.gradient_accumulation_steps,
+        num_training_steps=args.num_train_epochs * args.gradient_accumulation_steps,
+    )
+
+    # Prepare everything with our `accelerator`.
+    (
+        vae,
+        optimizer,
+        train_dataloader,
+        test_dataloader,
+        lr_scheduler,
+    ) = accelerator.prepare(
+        vae, optimizer, train_dataloader, test_dataloader, lr_scheduler
+    )
+
+    weight_dtype = torch.float32
+    if accelerator.mixed_precision == "fp16":
+        weight_dtype = torch.float16
+    elif accelerator.mixed_precision == "bf16":
+        weight_dtype = torch.bfloat16
+
+    vae.to(accelerator.device, dtype=weight_dtype)
+
+    # We need to initialize the trackers we use, and also store our configuration.
+    # The trackers initializes automatically on the main process.
+    if accelerator.is_main_process:
+        tracker_config = dict(vars(args))
+        accelerator.init_trackers(args.tracker_project_name, tracker_config)
+
+    num_update_steps_per_epoch = math.ceil(
+        len(train_dataloader) / args.gradient_accumulation_steps
+    )
+    args.max_train_steps = args.num_train_epochs * num_update_steps_per_epoch
+
+    # ------------------------------ TRAIN ------------------------------ #
+    total_batch_size = (
+        args.train_batch_size
+        * accelerator.num_processes
+        * args.gradient_accumulation_steps
+    )
+
+    logger.info("***** Running training *****")
+    logger.info(f"  Num examples = {len(train_dataset)}")
+    logger.info(f"  Num test samples = {len(test_dataset)}")
+    logger.info(f"  Num Epochs = {args.num_train_epochs}")
+    logger.info(f"  Instantaneous batch size per device = {args.train_batch_size}")
+    logger.info(
+        f"  Total train batch size (w. parallel, distributed & accumulation) = {total_batch_size}"
+    )
+    logger.info(f"  Gradient Accumulation steps = {args.gradient_accumulation_steps}")
+    global_step = 0
+    first_epoch = 0
+
+    progress_bar = tqdm(
+        range(global_step, args.max_train_steps),
+        disable=not accelerator.is_local_main_process,
+    )
+    progress_bar.set_description("Steps")
+
+    lpips_loss_fn = lpips.LPIPS(net="alex").to(accelerator.device)
+
+    for epoch in range(first_epoch, args.num_train_epochs):
+        vae.train()
+        train_loss = 0.0
+        for step, batch in enumerate(train_dataloader):
+            with accelerator.accumulate(vae):
+                target = batch["pixel_values"].to(weight_dtype)
+
+                # https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/autoencoder_kl.py
+                posterior = vae.encode(target).latent_dist
+                z = posterior.mode()
+                pred = vae.decode(z).sample
+
+                kl_loss = posterior.kl().mean()
+                mse_loss = F.mse_loss(pred, target, reduction="mean")
```

Comment:
In original stable-diffusion repo and SDXL repo, the vae loss is averaged over batch dim, which means they are summed in channel*height*width dims. Is this the correct way to average reconstruction loss?
https://github.com/CompVis/stable-diffusion/blob/21f890f9da3cfbeaba8e2ac3c425ee9e998d5229/ldm/modules/losses/contperceptual.py#L58